web/packages/teleterm/src/services/config/getKeyboardShortcutSchema.ts

```diff
@@ -0,0 +1,210 @@
+  return {
+    code: z.ZodIssueCode.custom,
+    message: `"${wrongKeyCode}" cannot be used as a key code.`,
+  };
+}
+
+export function invalidModifierIssue(wrongModifiers: string[]): z.IssueData {
+  return {
+    code: z.ZodIssueCode.custom,
+    message: `${wrongModifiers
+      .map(m => `"${m}"`)
+      .join(', ')} cannot be used as a modifier.`,
+  };
+}
+
+export function duplicateModifierIssue(): z.IssueData {
+  return {
+    code: z.ZodIssueCode.custom,
+    message: `Duplicate modifier found.`,
+  };
+}
+
+export function missingModifierIssue(keyCode: string): z.IssueData {
+  return {
+    code: z.ZodIssueCode.custom,
+    message: `"${keyCode}" requires a modifier.`,
+  };
+}
+
+export function getKeyboardShortcutSchema(platform: Platform) {
+  const allowedModifiers = getSupportedModifiers(platform);
+
+  return z
+    .string()
+    .transform(s => s.trim().split(/\s?\+\s?/))
+    .transform(putModifiersFirst(allowedModifiers))
+    .superRefine(validateKeyCodeAndModifiers(allowedModifiers))
+    .transform(adjustCasing)
+    .transform(s => s.join('+'));
+}
+
+function putModifiersFirst(
+  allowedModifiers: string[]
+): (tokens: string[]) => string[] {
+  return tokens =>
+    tokens.sort((a, b) => {
+      if (allowedModifiers.indexOf(a) === -1) {
+        return 1;
+      }
+      if (allowedModifiers.indexOf(b) === -1) {
+        return -1;
+      }
+      return allowedModifiers.indexOf(a) - allowedModifiers.indexOf(b);
+    });
+}
+
+/** Currently, works only for single characters.
+ * An accelerator string is compared to a string generated from event properties
+ * (so it should be case-insensitive too).
+ * But what is more important, this string is used also for tooltips.
+ * If we allow "CTRL+SHIFT+PAGEUP" then we cannot display it as "Ctrl+Shift+PageUp".
+ */
+function adjustCasing(tokens: string[]): string[] {
+  return tokens.map(token => {
+    if (token.length === 1) {
+      return token.toUpperCase();
+    }
+    return token;
+  });
+}
+
+function validateKeyCodeAndModifiers(
+  allowedModifiers: string[]
+): (tokens: string[], ctx: z.RefinementCtx) => void {
+  return (tokens, ctx) => {
+    // empty accelerator disables the shortcut
+    if (tokens.join('') === '') {
+      return z.NEVER;
+    }
+
+    const [keyCode, ...modifiers] = [...tokens].reverse();
+
+    const keyCodeUppercase =
+      keyCode.length === 1 // letters
+        ? keyCode.toUpperCase()
+        : keyCode;
+    if (!ALLOWED_KEY_CODES.includes(keyCodeUppercase)) {
+      ctx.addIssue(invalidKeyCodeIssue(keyCode));
+      return z.NEVER;
+    }
+
+    if (modifiers.length !== new Set(modifiers).size) {
+      ctx.addIssue(duplicateModifierIssue());
+      return z.NEVER;
+    }
+
+    const invalidModifiers = modifiers.filter(
+      modifier => !allowedModifiers.includes(modifier)
+    );
+    if (invalidModifiers.length) {
+      ctx.addIssue(invalidModifierIssue(invalidModifiers));
```

Comment:
The list of invalidModifiers should be run through a set, otherwise if you have duplicate invalid modifiers the error message says `"Command", "Command"`.

Reply:
Done